swarm/benches/connection_handler.rs

```diff
@@ -0,0 +1,284 @@
+        _peer_id: libp2p_identity::PeerId,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _event: libp2p_swarm::THandlerOutEvent<Self>,
+    ) {
+        self.spam_count = 0;
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p_swarm::ToSwarm<Self::ToSwarm, libp2p_swarm::THandlerInEvent<Self>>>
+    {
+        if self.spam_count == 0 {
+            std::task::Poll::Ready(libp2p_swarm::ToSwarm::GenerateEvent(FinishedSpamming))
+        } else {
+            std::task::Poll::Pending
+        }
+    }
+}
+
+struct PacketSpammerHandler {
+    spam_count: usize,
+    protocols: &'static [StreamProtocol],
+    outbound: Option<libp2p_swarm::Stream>,
+    inbound: Option<libp2p_swarm::Stream>,
+    stream_request_sent: bool,
+}
+
+impl ConnectionHandler for PacketSpammerHandler {
+    type FromBehaviour = Infallible;
+
+    type ToBehaviour = FinishedSpamming;
+
+    type InboundProtocol = Upgrade;
+
+    type OutboundProtocol = Upgrade;
+
+    type InboundOpenInfo = ();
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(
+        &self,
+    ) -> libp2p_swarm::SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        libp2p_swarm::SubstreamProtocol::new(Upgrade(self.protocols), ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<
+        libp2p_swarm::ConnectionHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::ToBehaviour,
+        >,
+    > {
+        let message = b"message";
+        let mut has_work_to_do = self.outbound.is_none() || self.inbound.is_none();
+
+        if let Some(outbound) = &mut self.outbound {
+            if self.spam_count > 0
+                && matches!(
+                    Pin::new(outbound).poll_write(cx, message),
+                    std::task::Poll::Ready(Ok(_))
+                )
+            {
+                self.spam_count -= 1;
+                has_work_to_do = true;
+            }
+        }
+
+        if let Some(inbound) = &mut self.inbound {
+            if matches!(
+                Pin::new(inbound).poll_read(cx, &mut [0; 1024]),
+                std::task::Poll::Ready(Ok(_))
+            ) {
+                has_work_to_do = true;
+            }
+        }
+
+        if !has_work_to_do {
+            std::task::Poll::Ready(libp2p_swarm::ConnectionHandlerEvent::NotifyBehaviour(
+                FinishedSpamming,
+            ))
+        } else if !self.stream_request_sent {
+            self.stream_request_sent = true;
+            std::task::Poll::Ready(
+                libp2p_swarm::ConnectionHandlerEvent::OutboundSubstreamRequest {
+                    protocol: libp2p_swarm::SubstreamProtocol::new(Upgrade(self.protocols), ()),
+                },
+            )
+        } else {
+            cx.waker().wake_by_ref();
+            std::task::Poll::Pending
+        }
```

Comment:
My suggestion would be to remove all of this and instead do as little work as we can in here. I think we might even be able to get away with doing nothing and just dropping all received streams?

What we want to measure is the overhead of the `Connection::poll` function itself and for that, the actual `ConnectionHandler` should do minimal work.

Here is my suggestion:

1. Instead of returning many protocols from 1 behaviour, make multiple benchmarks that all have one root behaviour with multiple composed ones (each perhaps with varying number of available protocols?). Because we compose behaviours into a tree, I'd expect there to be different performance characteristics when comparing 1 Behaviour with many protocols, vs many behaviour with one protocol each or many behaviours with many protocols.
2. In each behaviour, compose an instance of `libp2p::stream::Behaviour`. That one allows you to open streams to arbitrary protocols (meaning the one you use on the listener side doesn't need to support opening streams)
3. Ensure that you are not benchmarking the connection setup but only the part after.